### PR TITLE
Expand vulcan-exposed-ssh vulnerabilities and add labels

### PR DESCRIPTION
**Pull request by mm-sch**
- Adapt to the affected resources model
- Calculates the fingerprint of each vulnerability so actions applied to them can be dismissed if the context changes
- Adds labels to the vulnerabilities

### CODE REVIEW

File: cmd/vulcan-exposed-ssh/main.go [hunk 1]
```diff
@@ -176,10 +184,12 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
 			// NOTE: this doesn't cover different types of errors that ssh_scan can produce
 			continue
 		}
-		vulnDb := map[string]report.Vulnerability{}
 		// Exposed SSH
-		exposedSSHVuln.Details += fmt.Sprintf("* Exposed SSH Port in %v\n", i.Port)
-		vulnDb["exposedSSHVuln"] = exposedSSHVuln
+		v := exposedSSHVuln
+		v.Details += fmt.Sprintf("* Exposed SSH Port in %v\n", i.Port)
+		v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:190]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 2]
```diff
@@ -176,10 +184,12 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
+		v.Fingerprint = helpers.ComputeFingerprint(v.Summary, v.Score)
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:191]: If the summary changes, the entire issue will change so it's not useful for the fingerprint. I'd rather use or empty args, or the `ServerBanner` value. WDYT?

File: cmd/vulcan-exposed-ssh/main.go [hunk 3]
```diff
@@ -189,12 +199,15 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
+				ver, err := version.NewVersion(matches[1])
 				// Don't stop parsing the results if a version hasn't been found.
 				if err == nil {
-					if c6.Check(v) || c7.Check(v) || c8.Check(v) {
-						libsshVuln.Details += fmt.Sprintf("* libssh version %v in port %v may be vulnerable\n", v.String(), i.Port)
-						vulnDb["libsshVuln"] = libsshVuln
+					if c6.Check(ver) || c7.Check(ver) || c8.Check(ver) {
+						v := libsshVuln
+						v.Details += fmt.Sprintf("* libssh version %v in port %v may be vulnerable\n", ver.String(), i.Port)
+						v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:208]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 4]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
 			if strings.Contains(j, kexRecommendationPattern) {
-				tmp, ok := vulnDb["weakKexConfigVuln"]
-				if !ok {
-					tmp = weakKexConfigVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["weakKexConfigVuln"] = tmp
+				v := weakKexConfigVuln
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:228]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 5]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-				if !ok {
-					tmp = weakCiphersConfigVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["weakCiphersConfigVuln"] = tmp
+				v := weakCiphersConfigVuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:237]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 6]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-				if !ok {
-					tmp = weakMACsConfigVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["weakMACsConfigVuln"] = tmp
+				v := weakMACsConfigVuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:244]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 7]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-				if !ok {
-					tmp = passAuthVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["passAuthVuln"] = tmp
+				v := passAuthVuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:251]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 8]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-				if !ok {
-					tmp = allowSSHv1Vuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["allowSSHv1Vuln"] = tmp
+				v := allowSSHv1Vuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:258]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 9]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-				if !ok {
-					tmp = comprAlgoConfigVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["comprAlgoConfigVuln"] = tmp
+				v := comprAlgoConfigVuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:265]: I'd rather use something like `22/tcp` instead of `ssh`.

File: cmd/vulcan-exposed-ssh/main.go [hunk 10]
```diff
@@ -189,12 +199,15 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
 				// Don't stop parsing the results if a version hasn't been found.
 				if err == nil {
-					if c6.Check(v) || c7.Check(v) || c8.Check(v) {
-						libsshVuln.Details += fmt.Sprintf("* libssh version %v in port %v may be vulnerable\n", v.String(), i.Port)
-						vulnDb["libsshVuln"] = libsshVuln
+					if c6.Check(ver) || c7.Check(ver) || c8.Check(ver) {
+						v := libsshVuln
+						v.Details += fmt.Sprintf("* libssh version %v in port %v may be vulnerable\n", ver.String(), i.Port)
+						v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
+						v.Fingerprint = helpers.ComputeFingerprint(v.Summary, v.Score)
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:209]: If the summary changes, the entire issue will change so it's not useful for the fingerprint. I'd rather use or empty args, or the `ServerBanner` value. WDYT?

File: cmd/vulcan-exposed-ssh/main.go [hunk 11]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-					tmp = weakKexConfigVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["weakKexConfigVuln"] = tmp
+				v := weakKexConfigVuln
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.Fingerprint = helpers.ComputeFingerprint(v.Summary, v.Score)
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:231]: If the summary changes, the entire issue will change so it's not useful for the fingerprint. I'd rather use or empty args, or the `ServerBanner` value. WDYT?

File: cmd/vulcan-exposed-ssh/main.go [hunk 12]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-					tmp = weakCiphersConfigVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["weakCiphersConfigVuln"] = tmp
+				v := weakCiphersConfigVuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
+				v.Fingerprint = helpers.ComputeFingerprint(v.Summary, v.Score)
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:238]: If the summary changes, the entire issue will change so it's not useful for the fingerprint. I'd rather use or empty args, or the `ServerBanner` value. WDYT?

File: cmd/vulcan-exposed-ssh/main.go [hunk 13]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-					tmp = weakMACsConfigVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["weakMACsConfigVuln"] = tmp
+				v := weakMACsConfigVuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
+				v.Fingerprint = helpers.ComputeFingerprint(v.Summary, v.Score)
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:245]: If the summary changes, the entire issue will change so it's not useful for the fingerprint. I'd rather use or empty args, or the `ServerBanner` value. WDYT?

File: cmd/vulcan-exposed-ssh/main.go [hunk 14]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-					tmp = passAuthVuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["passAuthVuln"] = tmp
+				v := passAuthVuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
+				v.Fingerprint = helpers.ComputeFingerprint(v.Summary, v.Score)
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:252]: If the summary changes, the entire issue will change so it's not useful for the fingerprint. I'd rather use or empty args, or the `ServerBanner` value. WDYT?

File: cmd/vulcan-exposed-ssh/main.go [hunk 15]
```diff
@@ -211,62 +224,50 @@ func (r *runner) gradeVuln() ([]report.Vulnerability, error) {
...
-					tmp = allowSSHv1Vuln
-				}
-				tmp.Recommendations = append(tmp.Recommendations, j)
-				tmp.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
-				vulnDb["allowSSHv1Vuln"] = tmp
+				v := allowSSHv1Vuln
+				v.Recommendations = append(v.Recommendations, j)
+				v.Details += fmt.Sprintf("* Affected port: %v\n", i.Port)
+				v.AffectedResource = fmt.Sprintf("%d/%s", i.Port, "ssh")
+				v.Fingerprint = helpers.ComputeFingerprint(v.Summary, v.Score)
```
Comment (julianvilas) [inline cmd/vulcan-exposed-ssh/main.go:259]: If the summary changes, the entire issue will change so it's not useful for the fingerprint. I'd rather use or empty args, or the `ServerBanner` value. WDYT?

File: cmd/vulcan-exposed-ssh/main.go (hunk 1 above)
Comment (mm-sch) [reply cmd/vulcan-exposed-ssh/main.go:190]: Done!

File: cmd/vulcan-exposed-ssh/main.go (hunk 2 above)
Comment (mm-sch) [reply cmd/vulcan-exposed-ssh/main.go:191]: Yes, good point. I'll remove the args from the fingerprint call.

**Review by julianvilas — APPROVED** [review]
Perfect man, great job! :)

**Pull request merged**